Bio/SeqIO/__init__.py

```diff
@@ -844,8 +844,12 @@ def index(filename, format, alphabet=None, key_function=None):
 
     """
     # Try and give helpful error messages:
-    if not isinstance(filename, str):
-        raise TypeError("Need a filename (not a handle)")
+    if type(filename).__str__ is object.__str__:
+        raise TypeError("Need string for the filename (not a handle)")
+    elif not isinstance(filename, str):
+        # if filename is not a string, but can be converted to one, then
+        # convert it. Makes the API a bit nicer.
+        filename = str(filename)
```

Comment:
A better way is to assume that `filename` is valid, and check below if `proxy_class` raises an exception.

Reply:
Thanks for the review! I added a new commit that does this instead and something similar for `SeqIO.index_db`. This should also force the filename to be something that the `os` module accepts as a path, which makes a bit more sense.